include/openmc/plot.h

```diff
@@ -1,15 +1,204 @@
 #ifndef OPENMC_PLOT_H
 #define OPENMC_PLOT_H
 
+#include <unordered_map>
+#include <sstream>
+
+#include "xtensor/xarray.hpp"
+
 #include "hdf5.h"
+#include "openmc/position.h"
+#include "openmc/constants.h"
+#include "openmc/particle.h"
+#include "openmc/xml_interface.h"
 
 namespace openmc {
 
-extern "C" void voxel_init(hid_t file_id, const hsize_t* dims, hid_t* dspace,
+//===============================================================================
+// Global variables
+//===============================================================================
+
+constexpr int RED   = 0;
+constexpr int GREEN = 1;
+constexpr int BLUE  = 2;
+
+extern int PLOT_LEVEL_LOWEST; //!< lower bound on plot universe level
+
+extern std::unordered_map<int, int> plot_map; //!< map of plot ids to index
+
+extern "C" int32_t n_plots; //!< number of plots in openmc run
+
+class Plot;
+extern std::vector<Plot> plots; //!< Plot instance container
+
+//===============================================================================
+// RGBColor holds color information for plotted objects
+//===============================================================================
+
+struct RGBColor {
```

Comment:
This `struct` feels like a needless complication.  What does it provide that `std::array<unsigned char, 3>` does not?  The `std::array` also gives us a lot of nice features like an assignment operator and iterators.

Reply:
The main motivation was definitely the assignment operators. It cleans up the code quite well in the numerous places where we're reading `std::vector<int>`'s from the xml nodes.

The other issue it takes care of is literal definitions (see [previous definition](https://github.com/openmc-dev/openmc/blob/43488e02861c140fc9554dede7d9fda64cf6b298/src/plot.cpp#L40) of `WHITE`). This only happens once, but it's rather ugly to have to define a `const RGBColor` that way. Though now that its using `uint8_t` that may no longer be the case...

Reply:
`std::array` comes with an assignment operator.  For example:
```c++
#include <array>
#include <cstdint>
#include <iostream>

typedef std::array<uint8_t, 3> RGBColor;

int main() {
  RGBColor my_col {255, 255, 255};  // Can construct with an initializer-list
  my_col = {255, 100, 50};  // Can assign with an initializer-list or another array variable
  for (auto val : my_col) std::cout << static_cast<int>(val) << " ";
  std::cout << "\n";
  // This prints "255 100 50 "
}
```

And yeah, good call on the `uint8_t`.  I think that's definitely the type we want here.  It looks like it doesn't require those `static_cast`s like you showed before.

We can also avoid those annoying `reinterpret_cast`s by using the put-to operator:
```c++
  of.open(fname, std::ios::binary | std::ios::app);
  // Write color for each pixel
  for (int y = 0; y < pl.pixels_[1]; y++) {
    for (int x = 0; x < pl.pixels_[0]; x++) {
      for (auto val : rgb) of << val;
    }
  }
```

For dealing with those `std::vector<int>`s, just write a function that takes a `std::vector<int>` argument, checks the length, and returns a `std::array`

Reply:
:+1: agreed, this `struct` looks good without the `operator[]` definitions